### DISCUSSION TITLE
Not valid JSON while trying to use my API

### DISCUSSION THREAD

**Discussion by CathrinTruchan** [neuefische/questions#304] [Web] [answered]
Hello everyone, 

**Situation:**
I am trying to write an API to connect my app to my MongoDB. I have just set up a short API for GET and POST requests and a test-form just to see if it works. 

**The problem:**
When I try to submit the data in my test-form I get the following errors: 

`POST http://localhost:3000/api/flows 500 (Internal Server Error)` and 
`SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON`

I read, that this happens, when an API request responds with HTML instead of sending back JSON because of a server error. What can I do to fix this issue?

Here is my code for the API: 

```
export default async function handler(request, response) {
  if (request.method === "GET") {
    const flows = await getAllFlows();
    return response.status(200).json(flows);
  } else if (request.method === "POST") {
    await dbConnect();

   const postData = JSON.parse(request.body);
    const newFlow = await Flow.create(postData);

    return response
      .status(201)
      .json({ message: "Flow created", createdId: newFlow.id });
  }

  return response.status(405).json({ message: "HTTP method is not allowed" });
}

```
And here is my code from the test-form, where I am trying to submit the data: 

```
export default function testForm() {
  async function handleSubmit(data) {
    try {
      const response = await fetch("/api/flows", {
        method: "POST",
        body: JSON.stringify(data),
      });
      console.log(response);
      const result = await response.json();

      if (result.name) {
        alert("Flow has been created");
      } else {
        alert("Creating a flow did not work!!");
      }
    } catch (error) {
      console.error(error);
    }
  }

  function onSubmit(event) {
    event.preventDefault();
    const data = event.target.name.value;
    handleSubmit(data);
  }

  return (
    <form onSubmit={onSubmit} role="form">
      <label htmlFor="flow-name">Name:</label>
      <input
        aria-label="name"
        role="input"
        width="10rem"
        type="text"
        id="flow-name"
        name="name"
        minLength="5"
        maxLength="50"
        required
        autoFocus
      />

      <button aria-label="save" role="button">
        Save
      </button>
    </form>
  );
}

```

And here is the mongoose model, I am using:

```
const { Schema } = mongoose;

const flowSchema = new Schema({
  name: { type: String, required: true },
  description: { type: String },
  duration: {
    hours: { type: Number },
    minutes: { type: Number },
  },
  asanas: { type: Array },
});

const Flow =
  mongoose.models.Flow || mongoose.model("Flow", flowSchema, "flows");
 
```
Screenshot from Thunder Client: 

![Screenshot 2022-11-06 at 13 13 52](https://user-images.githubusercontent.com/111566392/200170231-753641cd-d9b4-44f4-8a55-6715658d5c2a.png)

I would really appreciate your help. Thank you!

**Comment by ThomasF85** [comment, marked as answer]
Hi Cathrin,

bei dir passiert ein Fehler in deiner API, irgendwo in diesem Stück Code:
```
} else if (request.method === "POST") {
    await dbConnect();

   const postData = JSON.parse(request.body);
    const newFlow = await Flow.create(postData);

    return response
      .status(201)
      .json({ message: "Flow created", createdId: newFlow.id });
  }
```
Schaue mal in die Konsole von VS Code, da siehst du Fehlermeldungen die auf dem Server auftreten. In der Browserkonsole fehlen dir diese Infos. Poste die Fehlermeldungen gerne nochmal hier - am besten mit einem Link zu einem aktuellen Commit.
Aus den Stücken Code die du hier gepostet hast werde ich nicht schlau. An einer Stelle sieht es so aus als würdest du nur einen String an die API schicken:
```
const data = event.target.name.value;
handleSubmit(data);
```
Das Objekt mit dem du mongoose fütterst und `Flow.create(object)` aufrufst muss die gleiche Struktur haben wie dein Model:
```
name: { type: String, required: true },
  description: { type: String },
  duration: {
    hours: { type: Number },
    minutes: { type: Number },
  },
  asanas: { type: Array },
```

**Reply by CathrinTruchan** [reply]
Hallo Thomas, 
danke für deine Antwort. Das Object was ich in der Form generiert habe, hatte nicht die gleiche Struktur wie das mongoose-model. Das habe ich geändert, jetzt bekomme ich im Browser nicht mehr die Fehlermeldung, dass JSON invalid ist, allerdings wird trotzdem kein neues Objekt erzeugt. Die Code stoppt jetzt an der Stelle "Creating a new flow did not work" 
```
export default function testForm() {
  async function handleSubmit(data) {
    try {
      const response = await fetch("/api/flows", {
        method: "POST",
        body: JSON.stringify(data),
      });
      console.log(data);
      const result = await response.json();
      console.log(result);
      if (result.name) {
        alert("Flow has been created");
      } else {
        alert("Creating a flow did not work!!");
      }
    } catch (error) {
      console.error(error);
    }
  }
```

Wenn ich die API im Thunder-Client teste und valides JSON übergebe, bekomme ich aber trotzdem noch den 500 internal server error. Hier die Fehlermeldung aus der VS Code Console: 

```
error - SyntaxError: Unexpected token o in JSON at position 1
    at JSON.parse (<anonymous>)
    at handler (webpack-internal:///(api)/./pages/api/flows/index.js:17:31)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.apiResolver (/Users/cathrintruchan/neuefische/capstone-project/node_modules/next/dist/server/api-utils/node.js:366:9)
    at async DevServer.runApi (/Users/cathrintruchan/neuefische/capstone-project/node_modules/next/dist/server/next-server.js:481:9)
    at async Object.fn (/Users/cathrintruchan/neuefische/capstone-project/node_modules/next/dist/server/next-server.js:735:37)
    at async Router.execute (/Users/cathrintruchan/neuefische/capstone-project/node_modules/next/dist/server/router.js:247:36)
    at async DevServer.run (/Users/cathrintruchan/neuefische/capstone-project/node_modules/next/dist/server/base-server.js:347:29)
    at async DevServer.run (/Users/cathrintruchan/neuefische/capstone-project/node_modules/next/dist/server/dev/next-dev-server.js:709:20)
    at async DevServer.handleRequest (/Users/cathrintruchan/neuefische/capstone-project/node_modules/next/dist/server/base-server.js:285:20) {
```

Hier ist der Link zu den Commits: 
https://github.com/CathrinTruchan/capstone-project/compare/main...feat/mongoDB-Connection

Ich bin mir nicht sicher, ob das jetzt zwei verschiedene Probleme sind?

**Reply by CathrinTruchan** [reply]
Ok, I am getting closer 😅 
I don't get any error messages anymore in the browser or in the console. The data are sent to the database and everything looks good. 
But I still get the error message, when I test the API POST request in the thunder client: 500 Internal Server Error. But since it works, I can maybe ignore that?

**Reply by ThomasF85** [reply]
I just tested your API and it works perfectly fine!
You have one problem: In the frontend in `test-form.js` you expect to get the newly created flow as a result and check for the result having the name attribute:
```
      if (result.name) {
        alert("Flow has been created");
      } else {
        alert("Creating a flow did not work!!");
      }
```
The Object you send from the API does not have the name attribute, because it looks like this:
```
{ message: "Flow created", createdId: newFlow.id }
```
you could for example change your API code to return the created flow:
```
    const newFlow = await Flow.create(postData);

    return response
      .status(201)
      .json({ message: "Flow created", flow: newFlow });
```
And check for the flow in the Frontend:
```
      if (result.flow) {
        alert("Flow has been created");
      } else {
        alert("Creating a flow did not work!!");
      }
```

**Reply by ThomasF85** [reply]
I guess I read your latest message too late and you already figured out what I just wrote. If there is a problem in the Thunder client but not in your Application, just ignore it!

**Reply by CathrinTruchan** [reply]
Hey Thomas, yes I already figured out, what the problem was :) But thanks for your help anyway!! 🙏
